dex/networks/eth/contracts/ETHSwapV1.sol

```diff
@@ -0,0 +1,231 @@
+    // has the following interpretation.
+    //   if (record == bytes32(0x00)): contract is uninitiated
+    //   else if (uint256(record) < block.number && sha256(record) != contract.secretHash):
+    //      contract is initiated and redeemable by the participant with the secret.
+    //   else if (sha256(record) == contract.secretHash): contract has been redeemed
+    //   else if (record == RefundRecord): contract has been refunded
+    //   else: invalid record. Should be impossible by construction
+    mapping(bytes32 => bytes32) public swaps;
+
+    // Vector is the information necessary for initialization and redemption
+    // or refund. The Vector itself is not stored on-chain. Instead, a key
+    // unique to the Vector is generated from the Vector data and keys
+    // the swap record.
+    struct Vector {
+        bytes32 secretHash;
+        uint256 value;
+        address initiator;
+        uint64 refundTimestamp;
+        address participant;
+    }
+
+    // contractKey generates a key hash which commits to the contract data. The
+    // generated hash is used as a key in the swaps map.
+    function contractKey(Vector calldata v) public pure returns (bytes32) {
+        return sha256(bytes.concat(v.secretHash, bytes20(v.initiator), bytes20(v.participant), bytes32(v.value), bytes8(v.refundTimestamp)));
+    }
+
+    // Redemption is the information necessary to redeem a Vector. Since we
+    // don't store the Vector itself, it must be provided as part of the
+    // redemption.
+    struct Redemption {
+        Vector v;
+        bytes32 secret;
+    }
+
+    function secretValidates(bytes32 secret, bytes32 secretHash) public pure returns (bool) {
+        return sha256(bytes.concat(secret)) == secretHash;
+    }
+
+    // constructor is empty. This contract has no connection to the original
+    // sender after deployed. It can only be interacted with by users
+    // initiating, redeeming, and refunding swaps.
+    constructor() {}
+
+    // senderIsOrigin ensures that this contract cannot be used by other
+    // contracts, which reduces possible attack vectors.
+    modifier senderIsOrigin() {
+        require(tx.origin == msg.sender, "sender != origin");
+        _;
+    }
+
+    // retrieveStatus retrieves the current swap record for the contract.
+    function retrieveStatus(Vector calldata v)
+        private view returns (bytes32, bytes32, uint256)
+    {
+        bytes32 k = contractKey(v);
+        bytes32 record = swaps[k];
+        return (k, record, uint256(record));
+    }
+
+    // status returns the current state of the swap.
+    function status(Vector calldata v)
+        public view returns(Status memory)
+    {
+        (, bytes32 record, uint256 blockNum) = retrieveStatus(v);
+        Status memory r;
+        if (blockNum == 0) {
+            r.step = Step.Empty;
+        } else if (record == RefundRecord) {
+            r.step = Step.Refunded;
```

Comment:
As per off-github convo, went with a check in `initiate` for the refund record hash, as well as a check server-side to prevent the maker choosing the refund record as the secret.